google/resource_compute_instance_migrate.go

```diff
@@ -152,3 +161,300 @@ func migrateStateV2toV3(is *terraform.InstanceState) (*terraform.InstanceState,
+			is.Attributes[fmt.Sprintf("attached_disk.%d.disk_encryption_key_sha256", attachedDisks)] = is.Attributes[fmt.Sprintf("disk.%d.disk_encryption_key_sha256", i)]
+
+			attachedDisks++
+		}
+	}
+
+	for k, _ := range is.Attributes {
+		if !strings.HasPrefix(k, "disk.") {
+			continue
+		}
+
+		delete(is.Attributes, k)
+	}
+	if scratchDisks > 0 {
+		is.Attributes["scratch_disk.#"] = strconv.Itoa(scratchDisks)
+	}
+	if attachedDisks > 0 {
+		is.Attributes["attached_disk.#"] = strconv.Itoa(attachedDisks)
+	}
+
+	log.Printf("[DEBUG] Attributes after migration: %#v", is.Attributes)
+	return is, nil
+}
+
+func getInstanceFromInstanceState(config *Config, is *terraform.InstanceState) (*compute.Instance, error) {
+	project, ok := is.Attributes["project"]
+	if !ok {
+		if config.Project == "" {
+			return nil, fmt.Errorf("could not determine 'project'")
+		} else {
+			project = config.Project
+		}
+	}
+
+	zone, ok := is.Attributes["zone"]
+	if !ok {
+		return nil, fmt.Errorf("could not determine 'zone'")
+	}
+
+	instance, err := config.clientCompute.Instances.Get(
+		project, zone, is.ID).Do()
+	if err != nil {
+		return nil, fmt.Errorf("error reading instance: %s", err)
+	}
+
+	return instance, nil
+}
+
+func getAllDisksFromInstanceState(config *Config, is *terraform.InstanceState) ([]*compute.Disk, error) {
+	project, ok := is.Attributes["project"]
+	if !ok {
+		if config.Project == "" {
+			return nil, fmt.Errorf("could not determine 'project'")
+		} else {
+			project = config.Project
+		}
+	}
+
+	zone, ok := is.Attributes["zone"]
+	if !ok {
+		return nil, fmt.Errorf("could not determine 'zone'")
+	}
+
+	diskList := []*compute.Disk{}
+	token := ""
+	for {
+		disks, err := config.clientCompute.Disks.List(project, zone).PageToken(token).Do()
+		if err != nil {
+			return nil, fmt.Errorf("error reading disks: %s", err)
+		}
+		diskList = append(diskList, disks.Items...)
+		token = disks.NextPageToken
+		if token == "" {
+			break
+		}
+	}
+
+	return diskList, nil
+}
+
+func getDiskFromAttributes(config *Config, instance *compute.Instance, allDisks map[string]*compute.Disk, attributes map[string]string, i int) (*compute.AttachedDisk, error) {
+	if diskSource := attributes[fmt.Sprintf("disk.%d.disk", i)]; diskSource != "" {
+		return getDiskFromSource(instance, diskSource)
+	}
+
+	if deviceName := attributes[fmt.Sprintf("disk.%d.device_name", i)]; deviceName != "" {
+		return getDiskFromDeviceName(instance, deviceName)
+	}
+
+	if encryptionKey := attributes[fmt.Sprintf("disk.%d.disk_encryption_key_raw", i)]; encryptionKey != "" {
+		return getDiskFromEncryptionKey(instance, encryptionKey)
+	}
+
+	autoDelete, err := strconv.ParseBool(attributes[fmt.Sprintf("disk.%d.auto_delete", i)])
+	if err != nil {
+		return nil, fmt.Errorf("error parsing auto_delete attribute of disk %d", i)
+	}
+	image := attributes[fmt.Sprintf("disk.%d.image", i)]
+
+	// We know project and zone are set because we used them to read the instance
```

Comment:
Do we though? From what I saw, the instance attributes were never updated with the project if we inferred the project from the config.

Reply:
Good point, fixed.